quickjs/src/main/java/com/squareup/quickjs/QuickJsException.java

```diff
@@ -18,10 +18,75 @@
 import androidx.annotation.Keep;
 import androidx.annotation.NonNull;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 @SuppressWarnings("unused") // Called from native code.
 @Keep // Instruct ProGuard not to strip this type.
 public final class QuickJsException extends RuntimeException {
+  /**
+   * QuickJs stack trace strings have multiple lines of the format "at func (file.ext:line)".
+   * "func" is optional, but we'll omit frames without a function, since it means the frame is in
+   * native code.
+   */
+  private final static Pattern STACK_TRACE_PATTERN =
+          Pattern.compile("\\s*at ([^\\s^\\[]+) \\(([^\\s]+):(\\d+)\\).*$");
+
+  /** Java StackTraceElements require a class name.  We don't have one in JS, so use this. */
+  private final static String STACK_TRACE_CLASS_NAME = "JavaScript";
+
   public QuickJsException(@NonNull String detailMessage) {
     super(detailMessage);
   }
+
+  public QuickJsException(@NonNull String detailMessage, @NonNull String jsStackTrace) {
+    super(detailMessage);
+    addJavaScriptStack(this, jsStackTrace);
+  }
+
+  /**
+   * Parses {@code StackTraceElement}s from {@code detailMessage} and adds them to the proper place
+   * in {@code throwable}'s stack trace.  Note: this method is also called from native code.
+   */
+  // TODO(szurbrigg): share this code with the DuktapeException implementation.
```

Comment:
I don’t think there’s _that_ much utility to sharing the code here?

Reply:
(in that it requires a new abstraction to share the code, and such an abstraction might be fragile)

Reply:
Yeah, probably not.  It started as a straight copy-paste, but it's already diverged a little.  Will delete the TODO in my next PR.